Artifactory: Added basic support to push binaries to an Artifactory store

andygrunwald opened the pull request:
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] I have read the **CONTRIBUTING** document.
- [ ] `make ci` passes on my machine.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.

### Comments

- `make ci` don't passes on my machine, because of failing `fpm` unit tests. It seems that this is not properly installed on my machine. Will try to make it work.
- Documentation: Only a first draft. Please read through and check if everything is understandable

## Description

This PR adds a new Pipeline to push the built binaries to the artifact story [Artifactory](https://www.jfrog.com/artifactory/) by JFrog.
This feature was requested in #344 and mentioned in #320.

## Notes

* How to push an artifact into Artifactory if described in their documentation: [Artifactory REST API: Example - Deploying an Artifact](https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#Artifactory%27sRESTAPI-DeployArtifact)
* As an example on how to build a new Pipeline, [draft: Docker integration #355](https://github.com/goreleaser/goreleaser/pull/355/files) was used
- Supported auth methods against Artifactory, only Basic Authentication with username/password or username/apikey is supported. See docs of Artifactory: [Access Tokens: Basic Authentication](https://www.jfrog.com/confluence/display/RTF/Access+Tokens#AccessTokens-BasicAuthentication)

## .goreleaser.yml

This PR extends the capabilities of ``.
Now it is possible to define multiple Artifactory instances like

```yaml
...
artifactories:
  - target: http://192.168.99.100:8081/artifactory/example-repo-local/goreleaser-tool/{{ .ProjectName }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}v{{ .Arm }}{{ end }}
    username: goreleaser
...
```

For every Artifactory, a secret (password or API key) has to be defined as well.
This is done like the `GITHUB_TOKEN` env var.

A new env var is introduced: `ARTIFACTORY_%d_SECRET`.

Means for the first configured artifactory, the password / api key will be stored in `ARTIFACTORY_0_SECRET`. For the second configured instance in `ARTIFACTORY_1_SECRET` and so on.

Both new keys (`target` and `username`) are mandatory if an entry in `artifactories` is set.
Defaults were not set in `pipeline/defaults/defaults.go`, because both keys are mandatory.

## How to test

You can start a Artifactory instance via Docker:

```sh
$ docker run --name artifactory -d -p 8081:8081 docker.bintray.io/jfrog/artifactory-oss:latest
```

If you have issues running this container, please check their documentation about this: [Installing Artifactory > Installing with Docker](https://www.jfrog.com/confluence/display/RTF/Installing+with+Docker).

I assume that your Docker Machine IP is 192.168.99.100 and this will be used in this test example.
Your instance is now available at http://192.168.99.100:8081/ 

To run first a manual test via curl, you can run

```sh
$ echo "This is a file" >> mybin
$ curl -u admin:admin -X PUT "http://192.168.99.100:8081/artifactory/example-repo-local/manual-test/mybin-v1.0.0/mybin.txt" -T ./mybin
```

In this request i assume that you chose `admin` as an username and `admin` as a password during the Artifactory start screen installation. When everything works you should see a response similar to this:

```json
{
  "repo" : "example-repo-local",
  "path" : "/manual-test/mybin-v1.0.0/mybin.txt",
  "created" : "2017-11-28T18:00:45.436Z",
  "createdBy" : "admin",
  "downloadUri" : "http://192.168.99.100:8081/artifactory/manual-test/mybin-v1.0.0/mybin.txt",
  "mimeType" : "application/octet-stream",
  "size" : "9",
  "checksums" : {
    "sha1" : "65d01857a69f14ade727fe1ceee0f52a264b6e57",
    "md5" : "a55e303e7327dc871a8e2a84f30b9983",
    "sha256" : "ead9b172aec5c24ca6c12e85a1e6fc48dd341d8fac38c5ba00a78881eabccf0e"
  },
  "originalChecksums" : {
    "sha256" : "ead9b172aec5c24ca6c12e85a1e6fc48dd341d8fac38c5ba00a78881eabccf0e"
  },
  "uri" : "http://192.168.99.100:8081/artifactory/example-repo-local/manual-test/mybin-v1.0.0/mybin.txt"
}
```

Now feel free to add this piece into a `.goreleaser.yml` of your choice

```yaml
artifactories:
  - target: http://192.168.99.100:8081/artifactory/example-repo-local/goreleaser-tool/{{ .ProjectName }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}v{{ .Arm }}{{ end }}
    username: goreleaser
```

and do a release.

When you are browsing into your Artifactory to your artifacts (file browser on the left) and you expand the tree, you will see that all your binaries are stored.

## Implementation comments

- Parts of the code of making the HTTP request + unit tests were inspired by [google/go-github](https://github.com/google/go-github)
- Is it "allowed" to import `"github.com/goreleaser/goreleaser/internal/buildtarget"` inside a pipeline?
- Personally i am not very happy about the solution to store the secrets in a format like `ARTIFACTORY_%d_SECRET`. Do you have a better idea? Maybe a `name` key per artifactory instance and then name the env vars like `ARTIFACTORY_<NAME>_SECRET`?

## Feedback

I would love to get feedback from you about the implementation. Maybe there are helper functions in the goreleaser code itself that we are able to use here.

GitCop commented:
There were the following issues with your Pull Request

 * Commit: 7f0247813843de21a6d38da9a0a497e4675c65c9
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

GitCop commented:
There were the following issues with your Pull Request

 * Commit: 7f0247813843de21a6d38da9a0a497e4675c65c9
  * Your commit message does not contain a body

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

andygrunwald commented:
Okay. After some back and forth, i fixed the unit tests, static code analyzer, added docs and added more unit tests (~82%). For now i will stop and wait for feedback.

caarlos0 commented on docs/160-artifactory.md:
```diff
@@ -0,0 +1,78 @@
+  -
+    # URL of your Artifactory instance + path to deploy to
+    target: http://artifacts.company.com:8081/artifactory/example-repo-local/{{ .ProjectName }}/{{ .Version }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}{{ .Arm }}{{ end }}
+    # User that will be used for the deployment
```
missing the next line here I think

andygrunwald replied:
Thanks. Missed it. Now there https://github.com/goreleaser/goreleaser/pull/429/files#diff-281972fdabc2eed834ff6e61de41a258R76

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,348 @@
+	sem := make(chan bool, ctx.Parallelism)
+	var g errgroup.Group
+
+	// Lets generate the build matrix, , because we want to publish
```
extra `,` here

andygrunwald replied:
Thx. Fixed.

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,348 @@
+
+// doBuild runs the pipe action of the current build and the current target
+// This is where the real action take place
+func doBuild(ctx *context.Context, build config.Build, target buildtarget.Target) (err error) {
```
maybe this function should be named `upload` or something? It is not really building here

andygrunwald replied:
You are right. Renamed it to `upload`.

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,348 @@
+
+		// The upload url to Artifactory needs the binary name
+		// Here we add the binary to the target url
+		if !strings.HasPrefix(uploadTarget, "/") {
```
I didn't get it...

you check for the prefix, and add a suffix `/`?

why?

andygrunwald replied:
You catched a bug. Fixed it to `HasSuffix`

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,348 @@
+		if err != nil {
+			return err
+		}
+		defer func() { err = file.Close() }()
```
this err here would be missed since the method return nil, right?

andygrunwald replied:
No, for this reason i introduced named returns in this function.
But i reverted it. The reason why i did this was because of gometalinter.
I ignored it now, because while fixing different gometalinter bugs, i introduced another bug which was fixed here. See https://github.com/goreleaser/goreleaser/pull/429/commits/37bc09f2bc4296ffb9d8e3e1e939529d8d4289a4

andygrunwald commented:
Thanks for the review @caarlos0 

> Other than that, I have a question about the overall implementation: shouldn't we upload the archives instead of the binaries? Or, maybe, allow this to be customised? Or is it an artifactory thing to upload binaries only?

To be honest: I don't know how other companies / people using Artifactory.
We use it with raw binaries. Maybe @skisulli from #334 can say something how he is using it?
But to make it configurable would make sense, i would say. Or even: It doesn't hurt anyone.

But @caarlos0, what is your opinion about the `ARTIFACTORY_%d_SECRET` env?
As mentioned above, i don't like the solution. What do you think? Would be a "name" identifier better? Any alternatives?

caarlos0 commented:
> To be honest: I don't know how other companies / people using Artifactory.
> We use it with raw binaries. Maybe @skisulli from #334 can say something how he is using it?
> But to make it configurable would make sense, i would say. Or even: It doesn't hurt anyone.

yeah, I never really used artifactory, so I don't know. let's see what @skisulli has to say. Will also ask on twitter just in case =)



> But @caarlos0, what is your opinion about the ARTIFACTORY_%d_SECRET env?
> As mentioned above, i don't like the solution. What do you think? Would be a "name" identifier better? Any alternatives?

yeah, it's not great, but works. Thinking about it now, a `name` identifier would also be useful on logging and debugging (instead of printing the full url or something, we can print the "artifactory name"). What do you think?

andygrunwald commented:
To be complete:
@caarlos0 sent out a tweet for feedback: https://twitter.com/caarlos0/status/937111190404259841
I also send out a twitter poll for feedback: https://twitter.com/andygrunwald/status/937254853843267584

## TODOs left

- [ ] Introduce name identifier per Artifactory instance and use this for logging + secret handling
- [ ] Raise Code coverage to > 92%
- [ ] Waiting for feedback from @skisulli and Twitter and see what happens
- [ ] Implement an "upload mode" per Artifactory instance: binaries only, archives only, both

I will see when i can implement those. My branch should be useable and editable for maintainers of goreleaser. If you are bored, feel free :)
Otherwise i will try to take care.

skisulli commented:
We currently upload the archives. We use goreleaser to build the binaries and archives and upload to GitHub. I’ve then written a small script to upload the archives to our Artifactory servers. I upload the archives so they match what is in GitHub.

andygrunwald commented:
@skisulli Can you share the script? Then i can validate if this PR covers the functionality of your script. And thanks for the feedback. I will implement an upload mode (like listed in the TODOs)

skisulli commented:
Here is the code from my Makefile

```make
upload:
	@cd dist && \
	for file in *.tar.gz ; do \
		echo "Uploading $$file"; \
		curl --progress-bar -m 60 -u"${ART_USERNAME}:${ART_PW}" -o /dev/null -T $$file "${ART_HOST}/artifactory/Apps/${APP}/${VERSION}/$$file"; \
	done && \
	echo "Uploading checksums.txt" && \
	curl --progress-bar -m 10 -u"${ART_USERNAME}:${ART_PW}" -o /dev/null -T ${APP}_$(subst v,,$(VERSION))_checksums.txt "${ART_HOST}/artifactory/Apps/${APP}/${VERSION}/${APP}-checksums.txt"

andygrunwald commented:
Just to let you know: I try to move this PR forward this weekend.

GitCop commented:
There were the following issues with your Pull Request

 * Commit: d5b97a2ef89291a98aa198322e101698e450f707
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

GitCop commented:
There were the following issues with your Pull Request

 * Commit: d5b97a2ef89291a98aa198322e101698e450f707
  * Your commit message body contains a line that is longer than 72 characters

Guidelines are available at https://github.com/goreleaser/goreleaser/blob/master/CONTRIBUTING.md#create-a-commit

---
This message was auto-generated by https://gitcop.com

andygrunwald commented:
Hey @caarlos0 and @skisulli,

finally i found some time to brush this up.
Changes since last time:

#### Introduced a `name` config per Artifactory

Every instance has now a name. Like

```yaml
artifactories:
  - name: production_eu
    target: http://192.168.99.100:8081/artifactory/example-repo-local/binary/{{ .ProjectName }}/{{ .Version }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}{{ .Arm }}{{ end }}
    username: admin
```

A name is required and used for the secret env var.
The env var for the configuration above is now `ARTIFACTORY_PRODUCTION_EU_SECRET` (Pattern: ARTIFACTORY_NAME-UPPERCASE_SECRET).

#### Upload mode

As the twitter survey said, people using the archives. Also thanks @skisulli for your make file.
You can now define an upload `mode` like

```yaml
artifactories:
  - name: production_binary
    mode: binary
    target: http://192.168.99.100:8081/artifactory/example-repo-local/binary/{{ .ProjectName }}/{{ .Version }}/{{ .Os }}/{{ .Arch }}{{ if .Arm }}{{ .Arm }}{{ end }}
    username: admin
  - name: production_archive
    target: http://192.168.99.100:8081/artifactory/example-repo-local/archive/{{ .ProjectName }}/{{ .Version }}/
    username: admin
```

Two upload modes are implemented:

* `binary`: Only uploads the raw binaries (like in the first version of this PR)
* `archive`: Uploads every generated artifact. This is also the default

**Important**: The variables _Os_, _Arch_ and _Arm_ are only available in `binary` mode. All other are available in both.

#### The test

When i execute the configuration from the *Upload mode*-chapter, goreleaser outputs

```
...
 • RELEASING TO ARTIFACTORY
      • skipped invalid build     target=darwin/arm6
      • skipped invalid build     target=darwin/arm64
      • uploaded successful       instance=production_binary mode=binary uri=http://192.168.99.100:8081/artifactory/example-repo-local/binary/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/Linux/arm6/goreleaser
      • uploaded successful       instance=production_binary mode=binary uri=http://192.168.99.100:8081/artifactory/example-repo-local/binary/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/Linux/arm64/goreleaser
      • uploaded successful       instance=production_binary mode=binary uri=http://192.168.99.100:8081/artifactory/example-repo-local/binary/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/Linux/i386/goreleaser
      • uploaded successful       instance=production_binary mode=binary uri=http://192.168.99.100:8081/artifactory/example-repo-local/binary/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/Linux/x86_64/goreleaser
      • uploaded successful       instance=production_binary mode=binary uri=http://192.168.99.100:8081/artifactory/example-repo-local/binary/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/Darwin/i386/goreleaser
      • uploaded successful       instance=production_binary mode=binary uri=http://192.168.99.100:8081/artifactory/example-repo-local/binary/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/Darwin/x86_64/goreleaser
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_Linux_armv6.tar.gz
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_Linux_arm64.tar.gz
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_Darwin_i386.tar.gz
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_Linux_i386.tar.gz
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_checksums.txt
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_Linux_x86_64.tar.gz
      • uploaded successful       instance=production_archive mode=archive uri=http://192.168.99.100:8081/artifactory/example-repo-local/archive/goreleaser/SNAPSHOT-240a70e17cb72cf5b88525384772ac508c35652c/goreleaser_Darwin_x86_64.tar.gz
...
```

And in my artifactory dummy instance, the files are uploaded: 
<img width="1433" alt="screen shot 2017-12-09 at 23 57 41" src="https://user-images.githubusercontent.com/320064/33800279-d36c6142-dd3c-11e7-95b9-0a96c9544c0d.png">

#### Merge checks

Only one check is failing. The check seems to expect a code coverage of 93.83% while i am hitting "only" 93.26%.

#### Feedback

@caarlos0 Please have a look at the code again. Feedback is welcome.
Otherwise i think this PR might be good to go.

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,429 @@
+
+// Default sets the pipe defaults
+func (Pipe) Default(ctx *context.Context) error {
+	if l := len(ctx.Config.Artifactories); l == 0 {
```
could be just `if len(ctx.Config.Artifactories) == 0 {`, right?

andygrunwald replied:
True. Will adjust it

andygrunwald replied:
Done in https://github.com/goreleaser/goreleaser/pull/429/commits/4e4529bbf8009ab762b0009766987d57ad5cc1ff

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,429 @@
+//
+// Docs: https://www.jfrog.com/confluence/display/RTF/Artifactory+REST+API#ArtifactoryRESTAPI-Example-DeployinganArtifact
+func (Pipe) Run(ctx *context.Context) error {
+	if l := len(ctx.Config.Artifactories); l == 0 {
```
could be simplified here as welll

andygrunwald replied:
Again the same as above :)

andygrunwald replied:
Done in https://github.com/goreleaser/goreleaser/pull/429/commits/4e4529bbf8009ab762b0009766987d57ad5cc1ff

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,429 @@
+
+	// Lets generate the build matrix, because we want
+	// to publish every target to Artifactory
+	for _, target := range buildtarget.All(build) {
```
I think you don't need to do that... you can iterate over the `ctx.Binaries` directly :)

andygrunwald replied:
I had a quick check and `ctx.Binaries` don't offer what the functionality represents.

`ctx.Binaries` looks like
```
map[string]map[string][]context.Binary{
    "darwinamd64":map[string][]context.Binary{
        "goreleaser_Darwin_x86_64":[]context.Binary{
            context.Binary{Name:"goreleaser", Path:"dist/goreleaser_Darwin_x86_64/goreleaser"}
        }
    }, 
    "linuxamd64":map[string][]context.Binary{
        "goreleaser_Linux_x86_64":[]context.Binary{
            context.Binary{Name:"goreleaser", Path:"dist/goreleaser_Linux_x86_64/goreleaser"}
        }
    },
...
```

and `buildtarget.All(build)` looks like
```
[]buildtarget.Target{
    buildtarget.Target{OS:"linux", Arch:"386", Arm:""}, 
    buildtarget.Target{OS:"linux", Arch:"amd64", Arm:""}, 
...
```

The reason for the `buildtarget.All(build)` usage is to be able to access the variables `OS`, `Arch` and `Arm` to replace them in the `target` template.
Those variables are not available in `ctx.Binaries`.

Do you share this? Or do i miss something?

Two options:
1. We keep the solution as is
2. We extend `ctx.Binaries` to extend the information

I prefer solution 1. for now (to get this functionality in and keep the PR "reviewable") and create a ticket for 2. to extend this feature and refactor this part here.
What do you think? I can create the tickets if you want.

caarlos0 replied:
Ohh, true, I agree. 

This binaries and artifacts things in the context package are pretty messed and I'll rewrite them eventually (#455)

let's leave it as-is for now :)

caarlos0 commented on pipeline/artifactory/artifactory.go:
```diff
@@ -0,0 +1,429 @@
+		}
+
+		envName := fmt.Sprintf("ARTIFACTORY_%s_SECRET", strings.ToUpper(instance.Name))
+		if os.Getenv(envName) == "" {
```
there is a `ctx.Env` map, maybe you could read from that.

it would also make testing easier I believe :)

andygrunwald replied:
Done in https://github.com/goreleaser/goreleaser/pull/429/commits/af618f9b690acd1b126010cc515eae5dc2f33c4c

andygrunwald commented:
> BTW: I think you need to merge with master as well

This PR is already sync with master.

andygrunwald commented:
Merged again with master + resolved conflict.

caarlos0 commented:
LGTM, let's ship this baby :D

andygrunwald commented:
This deserves a 🍺
